lite2/client.go

```diff
@@ -0,0 +1,419 @@
+		c.alternatives = providers
+	}
+}
+
+// Client represents a light client, connected to a single chain, which gets
+// headers from a primary provider, verifies them either sequentially or by
+// skipping some and stores them in a trusted store (usually, a local FS).
+//
+// Default verification: SkippingVerification(DefaultTrustLevel)
+type Client struct {
+	chainID          string
+	trustingPeriod   time.Duration // see TrustOptions.Period
+	verificationMode mode
+	trustLevel       cmn.Fraction
+
+	// Primary provider of new headers.
+	primary provider.Provider
+
+	// Alternative providers for checking the primary for misbehavior by
+	// comparing data.
+	alternatives []provider.Provider
+
+	// Where trusted headers are stored.
+	trustedStore    store.Store
+	trustedHeader   *types.SignedHeader // H
+	trustedNextVals *types.ValidatorSet // H+1
+
+	logger log.Logger
+}
+
+// NewClient returns a new light client. It returns an error if it fails to
+// obtain the header & vals from the primary or they are invalid (e.g. trust
+// hash does not match with the one from the header).
+//
+// See all Option(s) for the additional configuration.
+func NewClient(
+	chainID string,
+	trustOptions TrustOptions,
+	primary provider.Provider,
+	trustedStore store.Store,
+	options ...Option) (*Client, error) {
+
+	c := &Client{
+		chainID:          chainID,
+		trustingPeriod:   trustOptions.Period,
+		verificationMode: skipping,
+		trustLevel:       DefaultTrustLevel,
+		primary:          primary,
+		trustedStore:     trustedStore,
+		logger:           log.NewNopLogger(),
+	}
+
+	for _, o := range options {
+		o(c)
+	}
+
+	if err := c.initializeWithTrustOptions(trustOptions); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func (c *Client) initializeWithTrustOptions(options TrustOptions) error {
+	// 1) Fetch and verify the header.
+	h, err := c.primary.SignedHeader(options.Height)
+	if err != nil {
+		return err
+	}
+
+	// NOTE: Verify func will check if it's expired or not.
+	if err := h.ValidateBasic(c.chainID); err != nil {
+		return errors.Wrap(err, "ValidateBasic failed")
+	}
+
+	if !bytes.Equal(h.Hash(), options.Hash) {
+		return errors.Errorf("expected header's hash %X, but got %X", options.Hash, h.Hash())
+	}
+
+	// 2) Fetch and verify the next vals.
+	vals, err := c.primary.ValidatorSet(options.Height + 1)
+	if err != nil {
+		return err
+	}
+
+	// 3) Persist both of them and continue.
+	return c.updateTrustedHeaderAndVals(h, vals)
+}
+
+// SetLogger sets a logger.
+func (c *Client) SetLogger(l log.Logger) {
+	c.logger = l
+}
+
+// TrustedHeader returns a trusted header at the given height or nil if no such
+// header exist. It returns an error if there are some issues with the trusted
+// store, although that should not happen normally. TODO: mention how many
+// headers will be kept by the light client.
+//
+// 0 - the latest.
+// height must be >= 0.
+func (c *Client) TrustedHeader(height int64) (*types.SignedHeader, error) {
```

Comment:
Again, this trusted header may be outside the trusting period, correct?

Reply:
yes! added the check. thanks!